File: sunrise-backend-fastapi/app/schemas/alert.py
"""
Pydantic schemas for alert/notification system
Matches database schema in T800_alert_types.sql, T805_alert_statuses.sql, T810_alerts.sql
"""

from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime
from enum import Enum


# Enums for API validation
class AlertCategoryEnum(str, Enum):
    """Alert category options"""
    LEAVE_MANAGEMENT = "LEAVE_MANAGEMENT"
    FINANCIAL = "FINANCIAL"
    ACADEMIC = "ACADEMIC"
    ADMINISTRATIVE = "ADMINISTRATIVE"
    SYSTEM = "SYSTEM"


class AlertEntityTypeEnum(str, Enum):
    """Alert entity type options"""
    LEAVE_REQUEST = "LEAVE_REQUEST"
    FEE_PAYMENT = "FEE_PAYMENT"
    TRANSPORT_PAYMENT = "TRANSPORT_PAYMENT"
    STUDENT = "STUDENT"
    TEACHER = "TEACHER"
    ATTENDANCE = "ATTENDANCE"
    INVENTORY = "INVENTORY"


class ActorTypeEnum(str, Enum):
    """Actor type options"""
    STUDENT = "STUDENT"
    TEACHER = "TEACHER"
    ADMIN = "ADMIN"
    SYSTEM = "SYSTEM"


class TargetRoleEnum(str, Enum):
    """Target role options"""
    ADMIN = "ADMIN"
    TEACHER = "TEACHER"
    STUDENT = "STUDENT"


# Metadata response schemas
class AlertTypeResponse(BaseModel):
    """Alert type metadata response"""
    id: int
    name: str
    description: Optional[str] = None
    category: str
    icon: Optional[str] = None
    color_code: Optional[str] = None
    priority_level: int
    default_expiry_days: int
    requires_acknowledgment: bool
    is_active: bool

    class Config:
        from_attributes = True


class AlertStatusResponse(BaseModel):
    """Alert status metadata response"""
    id: int
    name: str
    description: Optional[str] = None
    color_code: Optional[str] = None
    is_final: bool
    is_active: bool

    class Config:
        from_attributes = True


# Alert schemas
class AlertBase(BaseModel):
    """Base schema for alerts"""
    title: str = Field(..., max_length=255, description="Alert title")
    message: str = Field(..., description="Alert message")
    entity_type: str = Field(..., max_length=50, description="Type of entity affected")
    entity_id: int = Field(..., description="ID of the affected entity")
    entity_display_name: Optional[str] = Field(None, max_length=255, description="Display name of entity")
    target_role: Optional[str] = Field(None, max_length=20, description="Target role: ADMIN, TEACHER, STUDENT")
    target_user_id: Optional[int] = Field(None, description="Specific target user ID")
    alert_metadata: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional context")
    priority_level: Optional[int] = Field(2, description="Priority: 1=Low, 2=Normal, 3=High, 4=Critical")


class AlertCreate(AlertBase):
    """Schema for creating alerts (internal use by services)"""
    alert_type_id: int = Field(..., description="Alert type ID from metadata")
    actor_user_id: Optional[int] = Field(None, description="User ID who triggered the action")
    actor_type: Optional[str] = Field(None, max_length=20, description="Actor type: STUDENT, TEACHER, ADMIN, SYSTEM")
    actor_name: Optional[str] = Field(None, max_length=200, description="Actor name for display")
    expires_at: Optional[datetime] = Field(None, description="Alert expiration datetime")


class AlertUpdate(BaseModel):
    """Schema for updating alert status"""
    alert_status_id: Optional[int] = Field(None, description="New status ID from metadata")


class AlertResponse(BaseModel):
    """Full alert response"""
    id: int
    alert_type_id: int
    alert_status_id: int
    title: str
    message: str
    entity_type: str
    entity_id: int
    entity_display_name: Optional[str] = None
    target_role: Optional[str] = None
    target_user_id: Optional[int] = None
    alert_metadata: Optional[Dict[str, Any]] = None
    priority_level: int
    actor_user_id: Optional[int] = None
    actor_type: Optional[str] = None
    actor_name: Optional[str] = None
    read_at: Optional[datetime] = None
    read_by: Optional[int] = None
    acknowledged_at: Optional[datetime] = None
    acknowledged_by: Optional[int] = None
    expires_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    # Enriched fields from relationships
    alert_type_name: Optional[str] = None
    alert_type_icon: Optional[str] = None
    alert_type_color: Optional[str] = None
    alert_type_category: Optional[str] = None
    alert_status_name: Optional[str] = None
    alert_status_color: Optional[str] = None

    class Config:
        from_attributes = True


class AlertWithDetails(AlertResponse):
    """Alert with full metadata details"""
    alert_type: Optional[AlertTypeResponse] = None
    alert_status: Optional[AlertStatusResponse] = None


class AlertListResponse(BaseModel):
    """Paginated alert list response"""
    alerts: List[AlertResponse]
    total: int
    page: int
    per_page: int
    total_pages: int
    unread_count: int


class AlertFilters(BaseModel):
    """Filter parameters for alert queries"""
    alert_type_id: Optional[int] = None
    alert_status_id: Optional[int] = None
    category: Optional[str] = None
    entity_type: Optional[str] = None
    target_role: Optional[str] = None
    priority_level: Optional[int] = None
    from_date: Optional[datetime] = None
    to_date: Optional[datetime] = None
    is_read: Optional[bool] = None


class AlertStats(BaseModel):
    """Alert statistics for dashboard"""
    total_alerts: int
    unread_count: int
    by_category: Dict[str, int]
    by_priority: Dict[str, int]
    recent_count: int  # Last 24 hours


class AlertUnreadCountResponse(BaseModel):
    """Unread count response"""
    unread_count: int


class AlertActionResponse(BaseModel):
    """Response for alert actions (mark read, acknowledge, etc.)"""
    message: str
    alert_id: Optional[int] = None
    count: Optional[int] = None

File: sunrise-school-frontend/src/types/alert.ts
/**
 * TypeScript types for Alert/Notification system
 * Matches backend schemas in sunrise-backend-fastapi/app/schemas/alert.py
 */

// Alert Type metadata
export interface AlertType {
  id: number;
  name: string;
  description?: string;
  category: string;
  icon?: string;
  color_code?: string;
  priority_level: number;
  default_expiry_days: number;
  requires_acknowledgment: boolean;
  is_active: boolean;
}

// Alert Status metadata
export interface AlertStatus {
  id: number;
  name: string;
  description?: string;
  color_code?: string;
  is_final: boolean;
  is_active: boolean;
}

// Main Alert interface
export interface Alert {
  id: number;
  alert_type_id: number;
  alert_status_id: number;
  title: string;
  message: string;
  entity_type: string;
  entity_id: number;
  entity_display_name?: string;
  target_role?: string;
  target_user_id?: number;
  alert_metadata?: Record<string, any>;
  priority_level: number;
  actor_user_id?: number;
  actor_type?: string;
  actor_name?: string;
  read_at?: string;
  read_by?: number;
  acknowledged_at?: string;
  acknowledged_by?: number;
  expires_at?: string;
  created_at: string;
  updated_at?: string;
  // Enriched fields from relationships
  alert_type_name?: string;
  alert_type_icon?: string;
  alert_type_color?: string;
  alert_type_category?: string;
  alert_status_name?: string;
  alert_status_color?: string;
}

// Alert with full metadata details
export interface AlertWithDetails extends Alert {
  alert_type?: AlertType;
  alert_status?: AlertStatus;
}

// Paginated alert list response
export interface AlertListResponse {
  alerts: Alert[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
  unread_count: number;
}

// Alert filters for queries
export interface AlertFilters {
  alert_type_id?: number;
  alert_status_id?: number;
  category?: string;
  entity_type?: string;
  target_role?: string;
  priority_level?: number;
  from_date?: string;
  to_date?: string;
  is_read?: boolean;
}

// Alert statistics for dashboard
export interface AlertStats {
  total_alerts: number;
  unread_count: number;
  by_category: Record<string, number>;
  by_priority: Record<string, number>;
  recent_count: number;
}

// Unread count response
export interface AlertUnreadCountResponse {
  unread_count: number;
}

// Action response (mark read, acknowledge, etc.)
export interface AlertActionResponse {
  message: string;
  alert_id?: number;
  count?: number;
}

// Alert category enum values
export type AlertCategory = 
  | 'LEAVE_MANAGEMENT'
  | 'FINANCIAL'
  | 'ACADEMIC'
  | 'ADMINISTRATIVE'
  | 'SYSTEM';

// Alert entity type enum values
export type AlertEntityType =
  | 'LEAVE_REQUEST'
  | 'FEE_PAYMENT'
  | 'TRANSPORT_PAYMENT'
  | 'COMBINED_PAYMENT'
  | 'STUDENT'
  | 'TEACHER'
  | 'ATTENDANCE'
  | 'INVENTORY';

// Alert status IDs (from database)
export const ALERT_STATUS = {
  UNREAD: 1,
  READ: 2,
  ACKNOWLEDGED: 3,
  DISMISSED: 4,
  EXPIRED: 5,
} as const;

// Priority levels
export const ALERT_PRIORITY = {
  LOW: 1,
  NORMAL: 2,
  HIGH: 3,
  CRITICAL: 4,
} as const;

